### PR TITLE
Refactor library and add WIP controller tests

### PR DESCRIPTION
**Pull request by ncooke3**
Refactored library to improve testability and clarity.

Key Changes:
- All of the heartbeat logging logic (whether or not to log) is now in `HeartbeatController`
- Moved `HeartbeatInfo`'s visibility to internal and decoupled it's internal dependencies.
    -  Motivation: `HeartbeatInfo` represents stored heartbeats. Its structure is optimized for storage requirements (limit on write) and differs significantly from the structure the heartbeats will be in when they are sent.
- Added a new public model called `HeartbeatsPayload` 
   - Allows type safe encoding into header string.
- Added instance management system to `HeartbeatStorage` (tests for `HeartbeatStorage` are a WIP)
  -  Motivation: Multiple instances of `HeartbeatStorage` with the same identifier should share the same instance.

### CODE REVIEW

**Comment by ncooke3** [comment]
CI is failing because the HeartbeatStorage tests are currently a WIP.

File: HeartbeatLogging/Sources/HeartbeatStorage.swift [hunk 1]
```diff
@@ -14,57 +14,84 @@
...
-/// A type that can store and remove hearbeats.
 protocol HeartbeatStorageProtocol {
-  func offer(_ heartbeat: Heartbeat)
-  func flush() -> HeartbeatInfo?
+  typealias HeartbeatInfoTransform = (HeartbeatInfo?) -> HeartbeatInfo?
+
+  func async(_ transform: @escaping HeartbeatInfoTransform)
+  // TODO: Evaluate if async variant of below API is needed.
+  func getAndReset(using transform: HeartbeatInfoTransform?) throws -> HeartbeatInfo?
 }
```
Comment (ncooke3) [inline HeartbeatLogging/Sources/HeartbeatStorage.swift:23]: Note to readers: In an effort to shift the "whether or not to log a heartbeat" logic to `HeartbeatController`, I needed to make simpler API for this class.

File: HeartbeatLogging/Sources/Weak.swift [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+
+// TODO: Document.
```
Comment (ncooke3) [inline HeartbeatLogging/Sources/Weak.swift:17]: Note to readers: See `HeartbeatStorage`'s Instance Management for how this weak ref box was used.

File: HeartbeatLogging/Sources/HeartbeatInfo.swift [hunk 3]
```diff
@@ -14,151 +14,108 @@
...
   }
-}
 
-private extension Date {
-  /// Calculates the time interval since a given `date`.
-  func timeIntervalSince(_ date: Date) -> TimeInterval {
-    timeIntervalSinceReferenceDate - date.timeIntervalSinceReferenceDate
+  func makeIterator() -> IndexingIterator<[Element]> {
+    circularQueue
+      .compactMap { $0 } // Remove `nil` elements.
```
Comment (paulb777) [inline HeartbeatLogging/Sources/HeartbeatInfo.swift:114]: Why are there nil elements?

File: HeartbeatLogging/Sources/HeartbeatStorage.swift [hunk 4]
```diff
@@ -14,57 +14,86 @@
 
 import Foundation
 
-/// A type that can store and remove hearbeats.
 protocol HeartbeatStorageProtocol {
-  func offer(_ heartbeat: Heartbeat)
-  func flush() -> HeartbeatInfo?
+  typealias HeartbeatInfoTransform = (HeartbeatInfo?) -> HeartbeatInfo?
+
+  func async(_ transform: @escaping HeartbeatInfoTransform)
```
Comment (paulb777) [inline HeartbeatLogging/Sources/HeartbeatStorage.swift:20]: Could be confusing to use a Swift keyword as a function name.  Is there something more descriptive?

**Review by paulb777 — APPROVED** [review]
A few questions, but nothing blocking. Good progress!

File: HeartbeatLogging/Sources/HeartbeatInfo.swift (hunk 3 above)
Comment (ncooke3) [reply HeartbeatLogging/Sources/HeartbeatInfo.swift:114]: The `circularQueue` is an array with a fixed size (Buffer is initialized with fixed capacity at initialization). Empty/placeholder array elements are `nil`. 

```swift
let fixedCapacityArray: Int? = Array(repeating: nil, count: 3)  // [nil, nil, nil]
```

File: HeartbeatLogging/Sources/Heartbeat.swift [hunk 5]
```diff
@@ -45,20 +47,20 @@ struct Heartbeat: Codable, Equatable {
...
   /// time periods, this property serves to capture all the time periods that the heartbeat can represent in
   /// a moving average.
-  var timePeriods: [TimePeriod] = []
+  let timePeriods: [TimePeriod]
 
   /// Intializes a `Heartbeat` with given `info` and, optionally, a `date` and `version`.
   /// - Parameters:
-  ///   - info: An anonymous piece of information to associate the heartbeat with.
+  ///   - agent: An anonymous piece of information to associate the heartbeat with.
   ///   - date: The date when the heartbeat was recorded. Defaults to the current date.
```
[the remark this answers is not in the document]
Comment (ncooke3) [reply HeartbeatLogging/Sources/Heartbeat.swift:55]: Yes, dates are based on device. Foundation's `Date` type represents [A specific point in time, independent of any calendar or time zone.](https://developer.apple.com/documentation/foundation/date#:~:text=a%20specific%20point%20in%20time%2C%20independent%20of%20any%20calendar%20or%20time%20zone.). Given that, I believe traveling across time zones won't matter as dates are compared independent of calendar/timezone. 

I think it's worth adding a unit test that tests for this behavior so I'll do that!

There is an edge case worth calling out though since dates are based on device. If device time is manually changed, the date returned from `Date()` will be distorted based on the actual time IRL. The current implementation is susceptible to this behavior as well though. It's another layer of complexity though for sure though.

File: HeartbeatLogging/Sources/HeartbeatStorage.swift [hunk 6]
```diff
@@ -14,57 +14,86 @@
...
     self.queue = queue ?? DispatchQueue(label: "com.heartbeat.storage.\(id)")
   }
 
-  func offer(_ heartbeat: Heartbeat) {
+  // MARK: - Instance Management
+
+  // TODO: Add tests for instance management.
+
+  /// <#Description#>
+  static var cachedInstances: [String: Weak<HeartbeatStorage>] = [:]
```
[the remark this answers is not in the document]
Comment (ncooke3) [reply HeartbeatLogging/Sources/HeartbeatStorage.swift:51]: We need a weak box because collections strongly retain values so cached `HeartbeatStorage` will never be deallocated otherwise.

File: HeartbeatLogging/Sources/HeartbeatController.swift [hunk 7]
```diff
@@ -41,17 +52,48 @@ public final class HeartbeatController {
   ///
   /// - Parameter info: A `String` identifier to associate a new heartbeat with.
   public func log(_ info: String) {
-    let newHeartbeat = Heartbeat(info: info)
-    storage.offer(newHeartbeat)
+    let (agent, date, capacity) = (info, dateProvider(), limit)
```
Comment (maksymmalyhin) [inline HeartbeatLogging/Sources/HeartbeatController.swift:55]: Optional. Is it same as just assigning values of the variables on separate lines like `let agent = info`? If so, using 3 separate lines for it looks more readable to me.

File: HeartbeatLogging/Sources/HeartbeatStorage.swift (hunk 4 above)
Comment (maksymmalyhin) [reply HeartbeatLogging/Sources/HeartbeatStorage.swift:20]: I like the more verbose `readAndWriteAsync` version so far as it saves the reader time since more information can be get from code without looking at API docs.

File: HeartbeatLogging/Sources/Storage.swift [hunk 8]
```diff
@@ -70,11 +56,23 @@ extension FileStorage: PersistentStorage {
...
   }
+
+  private func createDirectories(in url: URL) throws {
+    do {
+      try fileManager.createDirectory(
+        at: url,
+        withIntermediateDirectories: true,
+        attributes: nil
+      )
+    } catch CocoaError.fileWriteFileExists {
```
Comment (maksymmalyhin) [inline HeartbeatLogging/Sources/Storage.swift:67]: I like Swift error handling so much more than Objective-C one!

@ncooke3 @paulb777 I think improving Firebase error system to optimize for Swift handling can be a next significant API improvement.

File: HeartbeatLogging/Sources/HeartbeatStorage.swift [hunk 9]
```diff
@@ -14,57 +14,86 @@
...
-  func offer(_ heartbeat: Heartbeat) {
+  // MARK: - Instance Management
+
+  /// <#Description#>
+  private(set) static var cachedInstances: [String: Weak<HeartbeatStorage>] = [:]
+
+  /// <#Description#>
+  /// - Parameter id: <#id description#>
+  /// - Returns: <#description#>
+  static func getInstance(id: String) -> HeartbeatStorage {
```
Comment (maksymmalyhin) [inline HeartbeatLogging/Sources/HeartbeatStorage.swift:54]: FYI. We had a discussion with @VinayGuthal and @vkryachko yesterday and they mostly convinced me that a case when we have two Firebase app instances with the same app ID is very rare, so we can ignore it. If we don't account for this case then we can have an instance of storage Firebase app instance which may simplify the implementation (it was the case for Android). In this case storage would need to be keyed by the app ID and app name. No immediate implications for Apple SDK so far, but if implementation becomes too complex, feel free to apply this simplification as well.

File: HeartbeatLogging/Sources/Weak.swift [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+
+/// <#Description#>
+struct Weak<Object: AnyObject> {
```
Comment (maksymmalyhin) [inline HeartbeatLogging/Sources/Weak.swift:20]: The name looks too generic to me. I would prefer more verbose like `WeakContainer` or something along these lines.

File: HeartbeatLogging/Tests/Unit/HeartbeatControllerTests.swift [hunk 11]
```diff
@@ -15,45 +15,269 @@
...
-      return flushed
+    func getAndReset(using transform: HeartbeatInfoTransform?) throws -> HeartbeatInfo? {
+      let oldHeartbeatInfo = heartbeatInfo
+      heartbeatInfo = transform?(heartbeatInfo)
+      return oldHeartbeatInfo
     }
   }
+
+  /// Simulates the device system time.
+  class SystemClock {
```
Comment (maksymmalyhin) [inline HeartbeatLogging/Tests/Unit/HeartbeatControllerTests.swift:233]: This name associates with a real system clock implementation for me. I would suggest updating the name with more reference to its intended use, e.g. `TestClock`. Also it seems it could be just a private extension on `Date` type. WDYT?

File: HeartbeatLogging/Tests/Unit/HeartbeatControllerTests.swift [hunk 12]
```diff
@@ -15,45 +15,269 @@
...
+    )
+
+    assertHeartbeatControllerFlushesEmptyPayload(controller)
+
+    // When
+    // - Clock time 2021-11-01 @ 00:00:00 (EST)
+    controller.log("dummy_agent")
+
+    // - Advance to 2021-11-01 @ 23:59:59 (EST)
+    do { clock.advance(by: 60 * 60 * 24 - 1) }
```
Comment (maksymmalyhin) [inline HeartbeatLogging/Tests/Unit/HeartbeatControllerTests.swift:114]: I'm curious what is the motivation to wrap this statement into `do {}`?

File: HeartbeatLogging/Tests/Unit/HeartbeatControllerTests.swift [hunk 13]
```diff
@@ -15,45 +15,269 @@
...
+
+    var formattedDate: String {
+      HeartbeatsPayload.dateFormatter.string(from: date)
+    }
+  }
+
+  // TODO: - Revisit below assertion implementation.
+  // This can be simplified further by making HeartbeatsPayload conform to Equatable...
+  func assertEqualPayloadStrings(_ encoded: String, _ literal: String) {
+    let encodedData = Data(base64Encoded: encoded)!
```
Comment (maksymmalyhin) [inline HeartbeatLogging/Tests/Unit/HeartbeatControllerTests.swift:252]: Will using `try XCTUnwrap()` here and below make a potential failure easier to investigate in the future comparing to force unwrap?

File: HeartbeatLogging/Tests/Unit/HeartbeatStorageTests.swift [hunk 14]
```diff
@@ -17,178 +17,315 @@ import XCTest
...
+    XCTAssertNotNil(heartbeatStorage2)
+    XCTAssertEqual(
+      HeartbeatStorage.cachedInstances,
+      ["sparky": Weak(object: heartbeatStorage2)]
+    )
+    // When
+    heartbeatStorage2 = nil
+    // - Then
+    XCTAssertNil(heartbeatStorage2)
+    XCTAssertEqual(HeartbeatStorage.cachedInstances, [:])
```
Comment (maksymmalyhin) [inline HeartbeatLogging/Tests/Unit/HeartbeatStorageTests.swift:116]: Note that implementation details are used here which force the test to change when the implementation changes even without changing the public API. Asserting implementation details should be avoided when possible. Doesn't look like a big concern in this case, but it seems to be avoidable. Not a blocker for the PR.

**Review by maksymmalyhin — APPROVED** [review]
LGTM, a couple of non-blocking comments.

File: HeartbeatLogging/Sources/Weak.swift (hunk 10 above)
Comment (ncooke3) [reply HeartbeatLogging/Sources/Weak.swift:20]: `WeakContainer` is good. Would you prefer `WeakReference` or `WeakContainer`?

Comment (maksymmalyhin) [reply HeartbeatLogging/Sources/Weak.swift:20]: Both LGTM, but I like WeakContainer slightly more.

Comment (ncooke3) [reply HeartbeatLogging/Sources/Weak.swift:20]: SGTM, thanks

File: HeartbeatLogging/Sources/HeartbeatStorage.swift (hunk 4 above)
Comment (ncooke3) [reply HeartbeatLogging/Sources/HeartbeatStorage.swift:20]: Thanks @maksymmalyhin. I think I'll go with:
```swift
func readAndWriteAsync(using transform: @escaping HeartbeatInfoTransform)
```

File: HeartbeatLogging/Tests/Unit/HeartbeatControllerTests.swift (hunk 12 above)
Comment (ncooke3) [reply HeartbeatLogging/Tests/Unit/HeartbeatControllerTests.swift:114]: I was experimenting with trying to differentiate public API usage from test logic to get us in the desired state.  I was trying to emphasize that the instructions in the `do {}` block are not public API related operations but test related code to get us in the desired state.

I also didnt know you could use `do {}` blocks like that. It makes for a nicer syntax then doing `_ = { ... }()`!

BUT, from looking at this again, I think renaming `clock`  to `testClock` makes it clear enough without silly `do {}` stuff.

File: HeartbeatLogging/Tests/Unit/HeartbeatControllerTests.swift [hunk 15]
```diff
@@ -15,45 +15,248 @@
...
+    )
+
+    assertHeartbeatControllerFlushesEmptyPayload(controller)
+
+    // When
+    // - Clock time 2021-11-01 @ 00:00:00 (EST)
+    controller.log("dummy_agent")
+
+    // - Advance to 2021-11-01 @ 23:59:59 (EST)
+    testDate.addTimeInterval(60 * 60 * 24 - 1)
```
Comment (ncooke3) [inline HeartbeatLogging/Tests/Unit/HeartbeatControllerTests.swift:114]: Thanks to your comments @maksymmalyhin, I took a closer look and realized I was able to remove the `SystemTime` class entirely because the `HeartbeatController` initializer captures the `testDate` in an `@escaping` closure so we can treat the date with reference semantics (`testDate.addTimeInterval(60 * 60 * 24 - 1)`) and the controller's date provider will reflect those changes when we use it in public API implementation.

File: HeartbeatLogging/Tests/Unit/HeartbeatControllerTests.swift (hunk 12 above)
Comment (ncooke3) [reply HeartbeatLogging/Tests/Unit/HeartbeatControllerTests.swift]: Reworked in e9d0361

File: HeartbeatLogging/Tests/Unit/HeartbeatControllerTests.swift (hunk 11 above)
Comment (ncooke3) [reply HeartbeatLogging/Tests/Unit/HeartbeatControllerTests.swift]: Reworked in e9d0361. This led to some nice cleanup, see comment below.

File: HeartbeatLogging/Sources/HeartbeatController.swift (hunk 7 above)
Comment (ncooke3) [reply HeartbeatLogging/Sources/HeartbeatController.swift]: Reworked in e9d0361

File: HeartbeatLogging/Sources/Weak.swift (hunk 10 above)
Comment (ncooke3) [reply HeartbeatLogging/Sources/Weak.swift]: Reworked in e9d0361

File: HeartbeatLogging/Tests/Unit/HeartbeatStorageTests.swift (hunk 14 above)
Comment (ncooke3) [reply HeartbeatLogging/Tests/Unit/HeartbeatStorageTests.swift:116]: Hmm, thanks for pointing this out. I'd like to understand your point more. So in this case, are the exposed implementation details the fact that it is a `Dictionary<String:WeakContainer>`? 

So then `HeartbeatStorage.cachedInstances` should be private? 
Without checking backing dictionary, I'm not sure how to confirm that an instance is removed from `cachedInstances`... would checking that a new instance is returned be enough?

```swift
var heartbeatStorage: HeartbeatStorage? = HeartbeatStorage.getInstance(id: "sparky")
heartbeatStorage = nil

// Instead of inspecting dictionary contents, try to get Instance again and assert that it is not the same instance as before (implying that the old one was discarded and new one was created?).
let newHeartbeatStorage= HeartbeatStorage.getInstance(id: "sparky")
XCTAssert(newHeartbeatStorage !== heartbeatStorage) 
```

Comment (maksymmalyhin) [reply HeartbeatLogging/Tests/Unit/HeartbeatStorageTests.swift:116]: Yes, this looks sufficient to validate the API contract.

File: HeartbeatLogging/Tests/Unit/HeartbeatControllerTests.swift (hunk 13 above)
Comment (ncooke3) [reply HeartbeatLogging/Tests/Unit/HeartbeatControllerTests.swift]: Reworked in 48a668b
